libs/core/src/lib/button/button.component.ts

```diff
@@ -46,35 +57,58 @@ export class ButtonComponent extends AbstractFdNgxClass {
     @Input() size: string; // TODO: deprecated, leaving for backwards compatibility
 
     /** @hidden */
-    _setProperties() {
-        if (this.fdButtonClass) {
-            this._addClassToElement('fd-button');
-        }
-        if (this.compact) {
-            this._addClassToElement('fd-button--compact');
-        }
-        if (this.glyph) {
-            this._addClassToElement('sap-icon--' + this.glyph);
-        }
-        if (this.fdType) {
-            this._addClassToElement('fd-button--' + this.fdType);
-        }
-        if (this.options) {
-            if (typeof this.options === 'string') {
-                this._addClassToElement('fd-button--' + this.options);
-            } else if (Array.isArray(this.options)) {
-                this.options.forEach(option => {
-                    if (typeof option === 'string') {
-                        this._addClassToElement('fd-button--' + option);
-                    }
-                });
-            }
+    constructor(private _elementRef: ElementRef) {
+    }
+
+    /** Function runs when component is initialized
+     * function should build component css class
+     * function should build css style
+     */
+    ngOnInit(): void {
+        this.buildComponentCssClass();
+        this.buildComponentCssStyle();
+    }
+
+    @applyCssClass
+    /** CssClassBuilder interface implementation
+     * function must return single string
+     * function is responsible for order which css classes are applied
+     */
+    buildComponentCssClass(): string {
+        return [
+            this.fdButtonClass && 'fd-button',
+            this.compact && 'fd-button--compact',
+            this._glyph && `sap-icon--${this._glyph}`,
+            this.fdType && `fd-button--${this.fdType}`,
+            this.options && this.getOptionCssClass(this.options),
+            this._class
+        ].join(' ');
+    }
+
+    @applyCssStyle
+    /** CssStyleBuilder interface implementation
+     * function must return hashmap where
+     * key:string
+     * value:any
+     */
+    buildComponentCssStyle(): Hash {
+        return {
         }
     }
 
+    /** HasElementRef interface implementation
+     * function used by applyCssClass and applyCssStyle decorators 
+     */
+    elementRef(): ElementRef<any> {
+        return this._elementRef;
+    }
+
     /** @hidden */
-    constructor(private elementRef: ElementRef) {
-        super(elementRef);
+    private getOptionCssClass(options: string | ButtonOptions[]): string {
+        if (Array.isArray(this.options)) {
+            return this.options.reduce((_class, option) => _class += ` fd-button--${option}`, ' ')
```

Comment:
This will produce string with additional white spaces like: `"  fd-button--1 fd-button--2 fd-button--3"`. I'd recommend using `.map()` and `.join()`

Reply:
changes applied as suggested